File: perl/04-control-flow/control_flow_loops.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10;

# foreach iterates over each element of a list
my @colors = ("red", "green", "blue");
foreach my $color (@colors) {
    say "Color: $color";
}

# C-style for loop with an explicit counter
for (my $i = 1; $i <= 3; $i++) {
    say "Count: $i";
}

# The range operator (..) generates a list - great with foreach
for my $n (1..5) {
    print "$n ";
}
print "\n";

# while repeats while the condition is true
my $countdown = 3;
while ($countdown > 0) {
    say "T-minus $countdown";
    $countdown--;
}

# until repeats UNTIL the condition becomes true (opposite of while)
my $i = 0;
until ($i >= 2) {
    say "i is $i";
    $i++;
}

File: zig/04-control-flow/control_flow_loops.zig
const std = @import("std");

pub fn main() void {
    // while loop with a manual counter
    var i: u32 = 1;
    while (i <= 3) {
        std.debug.print("while: {d}\n", .{i});
        i += 1;
    }

    // while with a continue expression (runs after each iteration)
    var sum: u32 = 0;
    var n: u32 = 1;
    while (n <= 5) : (n += 1) {
        sum += n;
    }
    std.debug.print("sum 1..5 = {d}\n", .{sum});

    // for loop over an array of string slices
    const colors = [_][]const u8{ "red", "green", "blue" };
    for (colors) |color| {
        std.debug.print("color: {s}\n", .{color});
    }

    // for loop with an index using the 0.. range syntax
    for (colors, 0..) |color, index| {
        std.debug.print("{d} -> {s}\n", .{ index, color });
    }

    // for over an integer range
    for (0..3) |k| {
        std.debug.print("range: {d}\n", .{k});
    }
}
